### ISSUE TITLE
Recommendation for dasherizing (kebab-case-ing) namespaced object, such as `Admin::User`

### ISSUE THREAD

**Issue by bf4** [json-api/json-api#850]
JSON API recommends dasherized keys, so an `AdminUser` would be represented by `admin-user`. But, does/should JSON API have an opinion on namespaced objects? That `Admin::User` would be dasherized as `admin/user`?  (That's the result in Rails of `"Admin::User".underscore.dasherize`)

**Comment by joaomdmoura** [comment]
+1

**Comment by bf4** [comment]
I suppose this would be a PR to http://jsonapi.org/recommendations/#naming ?

**Comment by bf4** [comment]
Do I understand correctly from the spec that `SOLIDUS, "/"` is disallowed in a member name, so I'm thinking maybe it has to be `_`?

Spec recommendations http://jsonapi.org/recommendations/#naming

> Member names SHOULD contain only the characters "a-z", "0-9", and the hyphen minus  ("-") as separator between multiple words.

Spec Rules http://jsonapi.org/format/#document-member-names, basically:

> A-Z, a-z, 0-9, -, _,  SPACE (not recommended), any UNICODE character except U+0000 to U+007F (not recommended)
> With some disallowed characters http://jsonapi.org/format/#document-member-names-reserved-characters

**Comment by ethanresnick** [comment]
Yes, the solidus is not allowed and a hyphen, double hyphen, or underscore could be used instead. But I'm hesitant to make any sort of recommendation here, because, given the complexities around hierarchical types, I don't want clients to start inferring any meaning from whatever separator pattern we might recommend.

**Comment by bf4** [comment]
So, this Ember code not per JSON API?

https://github.com/emberjs/data/blob/aeccf38a5193a9fbf763a835588cc4ad315324b7/packages/ember-data/lib/serializers/json-api-serializer.js#L319-L351

>    By default `JSONAPISerializer` follows the format used on the examples of
>    http://jsonapi.org/format and uses dashes as the word separator in the JSON
>    attribute keys.
> 
> import { pluralize, singularize } from 'ember-inflector/lib/system/string';
> 
> var dasherize = Ember.String.dasherize;
> 
>   keyForAttribute: function(key, method) {
>     return dasherize(key);
>   },

Where dasherize is https://github.com/emberjs/ember.js/blob/v2.0.1/packages/ember-runtime/lib/system/string.js#L242-L258

>  Replaces underscores, spaces, or camelCase with dashes.
> 
>    'innerHTML'.dasherize();          // 'inner-html'
>     'action_name'.dasherize();        // 'action-name'
>     'css-class-name'.dasherize();     // 'css-class-name'
>     'my favorite items'.dasherize();  // 'my-favorite-items'
>     'privateDocs/ownerInvoice'.dasherize(); // 'private-docs/owner-invoice'

Where dasherize is defined as:

`str.replace(/([a-z\d])([A-Z])/g, '$1_$2').toLowerCase().replace((/[ _]/g, '-');`

> var STRING_DASHERIZE_REGEXP = (/[ _]/g);
> 
> var STRING_DASHERIZE_CACHE = new Cache(1000, function(key) {
>   return decamelize(key).replace(STRING_DASHERIZE_REGEXP, '-');
> });
> 
> var STRING_DECAMELIZE_REGEXP = (/([a-z\d])([A-Z])/g);
> 
> var DECAMELIZE_CACHE = new Cache(1000, function(str) {
>   return str.replace(STRING_DECAMELIZE_REGEXP, '$1_$2').toLowerCase();
> });

i.e. it leaves slashes alone

**Comment by ethanresnick** [comment]
Correct, that ember code would be non compliant; good catch there. I'd open an issue on the ember data repo.

**Comment by bf4** [comment]
Discussing in https://github.com/rails-api/active_model_serializers/issues/1216 `Admin::User` -> `admin--user` seems weird.  Any idea why `/` and `:` are disallowed? (or any non-alphanum, -, _, or ' ')

**Comment by youroff** [comment]
I think the main concern is that it's non url-safe. So someone would want to use resource type in url. So we could use `.` here, as it's url-safe and doesn't interfere with other word-separators.

Also we could use completely different approach by moving namespace to its own field next to `type` (and make it optional), or to the meta (which wouldn't require any standard changes, I assume).

**Comment by bf4** [comment]
Thanks @youroff but `.` isn't available either :( http://jsonapi.org/format/#document-member-names-reserved-characters https://github.com/json-api/json-api/blob/gh-pages/_format/1.0/normative-statements.json#L966-L1004

``` json
{
         "description": "it is **RECOMMENDED** that member names use only non-reserved, URL safe characters specified in RFC 3986.",
        "description": "The following \"globally allowed characters\" **MAY** be used anywhere in a member name:\\n\\n- U+0061 to U+007A, \"a-z\"\\n- U+0041 to U+005A, \"A-Z\"\\n- U+0030 to U+0039, \"0-9\"\\n- any UNICODE character except U+0000 to U+007F (not recommended, not URL safe)",
        "description": "The following characters **MUST NOT** be used in member names:\\n\\n- U+002B PLUS SIGN, \"+\" (used for ordering)\\n- U+002C COMMA, \",\" (used separator for multiple relationship paths)\\n- U+002E PERIOD, \".\" (used as relationship path separators)\\n- U+005B LEFT SQUARE BRACKET, \"[\" (use in sparse fieldsets)\\n- U+005D RIGHT SQUARE BRACKET, \"]\" (used in sparse fieldsets)\\n- U+0021 EXCLAMATION MARK, \"!\"\\n- U+0022 QUOTATION MARK, '\"'\\n- U+0023 NUMBER SIGN, \"#\"\\n- U+0024 DOLLAR SIGN, \"$\"\\n- U+0025 PERCENT SIGN, \"%\"\\n- U+0026 AMPERSAND, \"&\"\\n- U+0027 APOSTROPHE, \"'\"\\n- U+0028 LEFT PARENTHESIS, \"(\"\\n- U+0029 RIGHT PARENTHESIS, \")\"\\n- U+002A ASTERISK, \"*\"\\n- U+002F SOLIDUS, \"/\"\\n- U+003A COLON, \":\"\\n- U+003B SEMICOLON, \";\"\\n- U+003C LESS-THAN SIGN, \"<\"\\n- U+003D EQUALS SIGN, \"=\"\\n- U+003E GREATER-THAN SIGN, \">\"\\n- U+003F QUESTION MARK, \"?\"\\n- U+0040 COMMERCIAL AT, \"@\"\\n- U+005C REVERSE SOLIDUS, \"\\\"\\n- U+005E CIRCUMFLEX ACCENT, \"^\"\\n- U+0060 GRAVE ACCENT, \"`\"\\n- U+007B LEFT CURLY BRACKET, \"{\"\\n- U+007C VERTICAL LINE, \"|\"\\n- U+007D RIGHT CURLY BRACKET, \"}\"\\n- U+007E TILDE, \"~\""
}
```

**Comment by bf4** [comment]
Another answer might be `Admin::UserManager` -> `admin user-manager` and require the space be url encoded as `%20`?

In https://github.com/json-api/json-api/issues/567 I see that the intention of reserved is for future use, so maybe we could use `/`? 

sub-references: 
- https://github.com/json-api/json-api/pull/601 
- https://en.wikipedia.org/wiki/Naming_convention_%28programming%29#Multiple-word_identifiers
- https://tools.ietf.org/html/rfc3986
- https://github.com/json-api/json-api/issues/583
- https://github.com/kurko/ember-json-api/issues/53

cc people I think were involved in the discussion of member names @dgeb @ziege @tkellen @kurko

**Comment by youroff** [comment]
I know that it's not available by standard, I meant that if url-safety is a real concern, then `.` might be first candidate. Space is good option too, and it complies standard right now. There are plenty of good options actually, what can make things move is a decision of core maintainers of json-api.

**Comment by mtthgn** [comment]
If `/`is off the table, what about using two underscores (`__`)? This is how the ["Block, Element, Modifier"](http://cssguidelin.es/#bem-like-naming) CSS paradigm denotes hierarchy. 

This wouldn't conflict with people sending down multi-word keys with `_` in them (which is the current default for `ActiveModel::Serializer` with JSON API), and I think `admin__user` is a visually clear denotation of namespacing.

**Comment by fernandes** [comment]
hi guys,

I've thinking (researching) a little before posting here, and I'm in doubt why the situation below could not happen:

A compound model name in Ember/JS/CamelCase like `FooBar` is translated to `foo-bar`, so `AdminUser` should be `admin-user` and Ember would translate correctly. (this is how it works today)

About the namespace, considering the 3 options we have for [member names](http://jsonapi.org/format/#document-member-names)
- U+002D HYPHEN-MINUS, "-" (already used on compound name)
- U+005F LOW LINE, "_" (available one)
- U+0020 SPACE, " " (not recommended, not URL safe)

If we have something like Foo::BarBaz being translated into `foo_bar-baz` so JS would know how to handle namespace and compound name, as Ruby

My doubt is:

Considering the [recommendations](http://jsonapi.org/recommendations/), why just HYPHEN-MINUS is allowed and not LOW LINE? An according to [RFC3986](http://www.ietf.org/rfc/rfc3986.txt) LOW LINE is a unreserved character.

Recommendations came before the member names definition, so in my opinion, recommendations should be updated accordingly.

**TL;DR**
Why not add LOW LINE into recommendations and use `namespace_model-name` for name spaced model names

**Comment by ethanresnick** [comment]
I haven't discussed this issue with other spec editors, so I'm just speaking for myself, but: I'd like to reiterate that I'm hesitant to make any recommendation at all here on principle.

A namespace is either: 1) just a way of grouping multiple types, in which case the namespace is meta information (and can live in `meta` if it even needs to be exposed in the API at all); or 2) it's a sort of parent/hierarchical type, in which case [all kinds of subtleties](https://github.com/json-api/json-api/issues/862) apply and it's unlikely that trying to jam that hierarchy into the existing `type` attribute is a good idea.

So, my recommendation would be: 
- if you need the client to be able to group distinct `type`s, put the namespace in `meta` and leave it out of `type` all together;
- if you need to include the namespace solely to differentiate between the same type in different namespaces (e.g. `Admin::User` and `OtherNamespace::User`), then attach the namespace info _however you like_, because what matters to the client is not that a namespaces type is in use but just that different types have different `"type"`s.

**EDIT**: Actually, I want to soften my stance a bit. I can see the argument that the point of the dasherization recommendation is to help clients guess (and by default be right) what the type name will be that corresponds to a given class—and there's no reason we shouldn't extend that idea to cover classes that have a namespace. This argument makes perfect sense to me logically, but I guess I still don't see the need for namespaces as a very common case (i.e., why would/should an API's models live in different namespaces?) and I like the idea of type being more of an opaque string anyway. Therefore, if we were going to make a recommendation, I'd lean toward the least disruptive one, which would be the double hyphen (as in `admin--user`).

**Comment by fernandes** [comment]
hi @ethanresnick ,

Thank you for your reply, I checked the issues you referenced about hierarchical types

My concern is when you have, two models with the same `name` (not considering the namespace) and use `type` to differentiate both (because client needs to know they are different.

I'm trying to address what you told here

> if you need to include the namespace solely to differentiate between the same type in different namespaces (e.g. `Admin::User` and `OtherNamespace::User`), then attach the namespace info _however you like_, because what matters to the client is not that a namespaces type is in use but just that different types have different `"type"`s.

Exactly, we need to have different type, example, given two models:
- `Admin::Profile` (`name:string` `role:string`)
- `User::Profile` (`name:string` `document:string`)
- `User::AdditionalProfile` (`name:string`, `foo:string`, `baz:string`)

The client needs to know _which_ profile to instantiate with API data, if api provides just: `profiles` for both models, the client don't know how to handle the specific attributes `role` and `document`

If we use: `admin-profile` and `document-profile` its like `AdminProfile` and `DocumentProfile` and we are not making use of any namespace, but if we have possibility to provide this information on type, like:
- `admin_profile`
- `user_profile`
- `user_additional-profile`

And client like ember-data would know how to handle both types

(considering they _dasherize_ `privateDocs/ownerInvoice` to `private-docs_owner-invoice`)

its not about specialization like suggested in https://github.com/json-api/json-api/issues/571, its just about differentiate models using namespaces/`type`s.

Am I going into this issue right direction @bf4 ?

**Comment by bf4** [comment]
My gut based on discussion up till now is that whatever we're using the delimit words should be doubled for delimiting namespaces, or at least that's how I'm thinking of it in ruby and javascript
- `Admin:SuperUser` -> `admin__super_user` or `admin--super-user`

Ultimately, I'm thinking this will probably be a per-language convention, that is perhaps observed or noted but not 'recommended' by the JSON API.  And that convention will arise organically from what people do

**Comment by ethanresnick** [comment]
> Ultimately, I'm thinking this will probably be a per-language convention, that is perhaps observed or noted but not 'recommended' by the JSON API. And that convention will arise organically from what people do

:+1:

**Comment by ianzepp** [comment]
I'd like to chime in here as well, as I am dealing with the issue of namespaced JSON for a platform project.. To summarize the issue:
- Client installs a core object type from a public package (let's call it an `Account`).
- The Account object comes with a number of predefined properties, created by the package publisher. Each property needs to be logically in its own namespace. So for a property named `FirstName` we are looking at doing something like `Publisher:FirstName` as the JSON key name.
- On the client/subscriber side of things, the client needs to be able to extend the core object with their own properties (such as `MiddleName`). However, we need to ensure that the property name in the JSON object doesn't conflict with any property names from the original publisher. As a result, we are allowing clients to establish their own namespaces for their instances and/or to use a predefined global such as `Local` as the namespace.
- Finally, the namespacing needs to allow for additional future extensions to the object type from later package installs, that themselves may have extended the objects with unique (but potentially clashing) property names.

For us, the separator of choice seems to be a single colon, so we are ending up with JSON structures such as this:

```
{
  "id": "4jqhbsf3ej43bjfn",
  "core:description": null,
  "core:endpoint": "observers",
  "core:prefix": null,
  "core:suffix": null,
  "core:frozen": false,
  "core:stability": null,
  "core:restricted": false,
  "core:static": false,
  "core:immutable": false,
  "core:cacheable": true,
  "core:leaky": true,
  "core:metadata": true,
  "core:name": "Observer",
  "core:visibility": "global",
  "core-ui:icon": "fa-eye",
  "core-ui:label": "Observer",
  "core-ui:pluralLabel": "Observers",
  "db:namespaceId": "n0000000",
  "db:schemaId": "Schema",
  "db:createdDate": "2015-11-23T23:06:19.438Z",
  "db:createdById": "ekx4clbn",
  "db:updatedDate": "2015-12-09T22:23:55.565Z",
  "db:updatedById": "e1qnq2kvekxmex2j",
  "db:managedDate": "2015-11-23T23:06:19.438Z",
  "db:managedById": "ekx4clbn",
  "db:managed": true,
  "db:trashed": false,
  "db:deleted": false,
  "db:published": false,
  "core-sharing:read": "*",
  "core-sharing:edit": "*",
  "core-sharing:hide": "",
  "meta:editable": true
}
```

In this example, `core`, `core-ui`, and `core-sharing` are all packages which may or may not be installed, based on the subscriber's choice. The `db:*` package wraps the underlying database metadata, and the `meta:*` package is on-the-fly calculations of runtime-scope-specific metadata.

**Comment by ethanresnick** [comment]
@ianzepp We're having an _extensive_ discussion about this in #915. Please feel free to chime in there. For now, though, note that using a colon is [currently illegal](http://jsonapi.org/format/#document-member-names-reserved-characters) in a key name—precisely because it's reserved to allow json:api to add new features, like the namespacing being discussed in #915.

**Comment by ianzepp** [comment]
@ethanresnick Had a look over there, and think I'll stay out of that discussion. I didn't like the idea of trying to force preregistration of all namespaces (and require the use of a mandatory jsonapi.org "extension" profile) just so clients can process the payload itself via an `Accept` header. It also seems that most of the examples over there imply the use of a colon as a separator anyways.. e.g. #915:

> The implementation is simple. Keys can be used like base spec keys, except that the namespace prefix has to be added. Some languages might have problems with the used namespace separator, but this topic has been discussed before and is not relevant for the spec. Every language should be able to set valid JSON keys.

I also feel part of this is the fact that I'm not currently a fan of the JSON API approach anyways (via the example here: http://jsonapi.org/):
- too deep a data structure causes issues with null-pointers and/or requirements to use a library like lodash/underscore for _.get/_.set across nested objects which may or may not be present,
- the `link` and `href` entities assume a website type of relationship instead of an API focus (public URLs are irrelevant for services)
- the data structures are fragile and non-persistable in a document store if the front-end URLs ever change (assuming link/href is included)
- too much duplication of entities.. I don't need to include the author 500 times in a document for every comment, 
- and finally, once I've stripped out all the above features, the resulting payload is essentially flat and back to a normal object structure.

Mostly, though, I just wanted to chime in on the namespace separator thing. :)

Edit: to clarify bullet 2 above.. I probably spoke too strongly with the use of "irrelevant". I just don't feel permanent links (whether to API endpoints to public facing URLs) are appropriate data in a JSON payload, but instead should be built into the application itself. If I'm accessing data from example.org, I don't need to know in every payload that `comments` can be found at `example.com/api/v1/comments/:id`. That information is built into the functionality itself.

**Comment by bintoro** [comment]
@ianzepp Interesting. I can't say I recognize much of that criticism.

> too deep a data structure causes issues with null-pointers and/or requirements to use a library like lodash/underscore for .get/.set across nested objects which may or may not be present

Yeah I guess, if you're dealing with a raw payload directly without turning the representation into proper objects first. I've never run into this issue.

> the link and href entities assume a website type of relationship instead of an API focus (public URLs are irrelevant for services)

URLs are not required.

> the data structures are fragile and non-persistable in a document store if the front-end URLs ever change (assuming link/href is included)

Same as above.

> too much duplication of entities.. I don't need to include the author 500 times in a document for every comment,

Inclusion of every field/relationship is not required, and, independent of that, neither is inclusion of any related resources. Related resources are also never duplicated when they _are_ included.

**Comment by ethanresnick** [comment]
> I didn't like the idea of trying to force preregistration of all namespaces (and require the use of a mandatory jsonapi.org "extension" profile) just so clients can process the payload itself via an Accept header.

@ianzepp Re-reading your original post, I'm realizing that #915 is dealing with a more specific context than what you're talking about and involves a lot of background requirements (e.g. around registration) that may not make sense for your use case.

For what you're doing, all that matters is this: JSON:API may (soon) start using the colon to support some form of namespacing that works differently than the namespacing used in your app. If you don't generic JSON:API clients consuming your API to get confused when they see your app's use of `:`, make sure your API uses a different separator character (like the double hyphen, double underscore, etc). More generally, make sure your API avoids in key names all the characters that JSON:API has [reserved](http://jsonapi.org/format/#document-member-names-reserved-characters), and you won't run into any trouble.

As far as your other criticisms go, feel free to post them on [our forum](http://discuss.jsonapi.org/), where I and others can address them. Let's not have those broader discussions in this issue, though, since I think they're a bit out of scope.

**Comment by bintoro** [comment]
There's a clear need to have at least _some_ character that could be used for namespacing attributes, and I think the colon is a pretty obvious candidate. I don't get why having to URL encode it is such a big deal. It's not like people would be _forced_ to namespace their attributes.

**Comment by ethanresnick** [comment]
> There's a clear need to have at least some character that could be used for namespacing attributes, and I think the colon is a pretty obvious candidate. 

@bintoro I'm gonna respond to you in #915 shortly, but just to this issue of the colon... The issue isn't that it needs to be URI encoded; it's about what a key like `"a:b"` should denote. It could denote:
1. Nothing more than an application-specific group and sub-group, as in `user:profile`. Here, it's the application's job to break the values up into groups (like `user`) but, between different applications, there's no global mechanism for identifying a particular group. So, here, `db:core` in @ianzepp's application could mean something very different in mine.
2. A {namespace URI, local name} pair, as in xml. So, `db` maps to a URI and then my `db:core` is known to be the same as someone else's, if their prefix maps to the same URI.
3. A single URI (as in CURIEs, and my proposal for link relations in #915). So `db:core` means "the URI formed by appending the string 'core' to the URI for db".

The first type of namespacing is, by my read, what this issue's been about. But, in that version of namespacing, the whole key name is already inherently application specific, so I don't see a reason to unreserve the `:` (or anything else) when an application can repurpose an existing unreserved character. Am I missing something?

Therefore, if JSON API were going to specify some form for namespacing (which has only been needed for extensions so far), I'd think it would be to enable the second or third form. Now, the third form is arguably simpler conceptually, and it's definitely the favorite of (admittedly URI-obssessed) formats like RDF. So JSON API might implement version 3. Then, since the `a:b` keys in version 3 simply expand to a URI, we'd probably want to allow a URI to be used directly in all the places where `a:b` could be used. And that's where the problem with the colon comes in: if you use a colon, `a:b` could be either a URI (like `mailto:john@example.com` or `urn:.....`) or a namespaced value. So then you need to invent some way to distinguish between the two (hence Safe CURIEs)... but all that madness can be avoided if you just use something other than the `:` as the separator so that there's no ambiguity between a URI and the concise form.

**Comment by bintoro** [comment]
> I don't see a reason to unreserve the `:` (or anything else) when an application can repurpose an existing unreserved character.

Existing unreserved character such as an emoji? The problem is that all common delimiter characters _are_ reserved, which is highly annoying.

Releasing one character for the opaque, app-specific namespacing variant doesn't preclude the spec from implementing generic namespacing later using some other character.

**Comment by ethanresnick** [comment]
> Releasing one character for the opaque, app-specific namespacing variant doesn't preclude the spec from implementing generic namespacing later using some other character.

Fair enough! Which would you release? And why that one?

EDIT: And shouldn't we figure out #915 before releasing anything?

**Comment by bintoro** [comment]
> Fair enough! Which would you release? And why that one?

Anything that doesn't look outright horrible. The colon aside, any of `/ | >` would probably fit the bill.

But this got me thinking. Why not just use the colon for all our namespacing needs? There are at least two ways to do this:
1. Use `::` for opaque identifiers and `:` for CURIEs or whatever.
2. Use `:` for both kinds. This only requires that we disallow full URIs in keys, which is not very practical anyway. Parsing is quite straightforward: if the prefix is not in the document's prefix-to-URI mapping, it's not a URI.

---

**EDIT:** If it's desirable to allow full URIs everywhere, then the CURIE plan will likely use something else, like a slash, meaning that there should be no obstacle in assigning `:` for private namespacing purposes. After all, the potential confusion between a `namespace:localname` construct and a URI is a non-issue for the opaque identifiers which are not intended to be portable in that way.

**Comment by ethanresnick** [comment]
Ok, I'm gonna bring this up with the other editors at our next meeting, and we'll try to settle on a solution to it and #898 at the same time. (I mention #898 because, if we want to allow relative URIs in some places, then things could get more complicated, as a `/` isn't sufficient to distinguish a CURIE from a relative URI. But we'll see whether all this complexity is necessary.)

**Comment by bintoro** [comment]
> I mention #898 because, if we want to allow relative URIs in some places, then things could get more complicated, as a `/` isn't sufficient to distinguish a CURIE from a relative URI.

...which is yet another reason to just stick with the standard CURIE notation and use the colon.

Keys:
- begins with `prefix:` with mapped prefix ➔ CURIE
- begins with `prefix:` with unmapped prefix ➔ private namespace, treat as plain identifier
- begins with `prefix://` ➔ absolute URL
- everything else ➔ plain identifier

URL values:
- begins with `prefix:` with mapped prefix ➔ CURIE
- begins with `prefix://` ➔ absolute URL
- everything else ➔ relative URL

The only apparent limitation of this system is the one I mentioned before: keys may not contain full URIs that are not HTTP URLs, but that seems a rather academic concern.

---

**EDIT:**

To generalize from URL values to URIs:
- begins with `prefix:` with mapped prefix ➔ CURIE
- begins with `prefix:` with unmapped prefix ➔ absolute URI
- everything else ➔ relative URI

The only limitation with this is that a relative reference that looks like a URI must either encode the colon or use a root-relative path. That is, linking from `/api/foo/` to `/api/foo/bar:xx:yy` is not possible using the unencoded segment `bar:xx:yy`. But this is a pretty reasonable restriction, since such a link wouldn't work in a browser either.

**Comment by ethanresnick** [comment]
There are a few other complications (e.g. registered link relations could get confused with relative URIs), but I hear/understand what you're saying. 

I can't be online much over the next couple days, but I'm gonna try to assimilate all the discussion here and in #915 into a concrete proposal that can hopefully be ready in about a week.

**Comment by bintoro** [comment]
Alright :) Looking forward to it.

The relative URL vs. registered relation overlap exists regardless of the namespacing question and is limited to links' **keys** only, not their values.

It's pretty clear that you can't put relative URLs in keys because we'd also have to disambiguate between `self` and a relative link to an endpoint called `"self"`. Having a static list of names to compare against would make it impossible to add new reserved keys compatibly. More to the point, the use case for relative URLs in keys is likely nonexistent.

**Comment by mattxyzeth** [comment]
Just wondering what the outcome of that meeting discussion was. Was there any resolution on how to represent namespaces in type?

**Comment by mattxyzeth** [comment]
I just want to say that I'm personally not opposed to the idea of having the namespace defined in the meta. I really does feel like meta information, and it also allows for customizing the value so the client and the server would not have to share the exact same paths.

**Comment by mattxyzeth** [comment]
I guess the meta definition would need to be an object of key values though, to take into consideration side loaded resources in the included array.

Something like:

``` json
"meta": {
  "namespaces": {
    "resource-type": "namespace-value"
  }
}
```

**Comment by bintoro** [comment]
@mattgrannary Realistically, if someone needs namespaced attributes, it's quite likely they'll just include the namespace in the attribute name. Looking it up in `meta` requires you to control the client, and if you can control the client, you might as well use a reserved character.

Besides, using a reserved character in attribute names seems future-proof because the specification has painted itself into a corner with its compatibility promise. Since clients are required to ignore unrecognized members, I don't see how the spec could ever assign a special meaning to any character that is currently reserved. A generic server could not safely use the feature because it wouldn't know if the client just silently discarded those attributes.

**Comment by mattxyzeth** [comment]
I guess it is a bit needless to introduce a whole new layer of spec just for namespaces. Ultimately it would be simple and the most useful to just allow the solidus character in the type value. @youroff mentioned:

> I think the main concern is that it's non url-safe. So someone would want to use resource type in url. So we could use . here, as it's url-safe and doesn't interfere with other word-separators.

This confuses me because the solidus character is url-safe and if the resource type is used in the url separating the namespace and resource with a solidus, doesn't that make sense in terms on the actual resource location? By very definition of the [spec](http://www.ietf.org/rfc/rfc3986.txt) in section 1.2.3 this is how resource hierarchal namespacing works.

**Comment by bintoro** [comment]
The disallowed characters aren't disallowed due to URL safety concerns but because the specification authors are reserving them for future enhancements.

Meanwhile, developers can't use a private namespace system just in case the spec some day comes up with its own, which no one has even asked for. That's why this particular question is so ridiculous. The issue shouldn't even exist to begin with.

The project should choose its battles. I'm concerned it's becoming paralyzed trying to cover too many angles with too few resources.

**Comment by mattxyzeth** [comment]
True. I would argue that this would be one of these future enhancements.

**Comment by dijonkitchen** [comment]
Should this be closed? 

It's available via ActiveModel Serializers: https://github.com/rails-api/active_model_serializers/blob/master/docs/general/key_transforms.md

**Comment by kurko** [comment]
Disclaimer: I have not seem a concrete example in this thread to justify a _namespace_ character.

1. From a data design perspective, `user` and `admin-user` are the *same* thing/entity/resource. You're trying to bend JSONAPI to fit your needs because you have a flaw in your design. The only difference between those 2 is their roles, therefore what would change between them would probably be just their `role` attribute. For instance, `{"user": {"role": "admin"}` and `{"user": {"role": "visitor"}`, or we could perhaps go a step forward and add an object attribute, `{"user": {"role": "admin", "admin-details": {}}`.

2. JSONAPI is about data and ideally should not expose code modules/namespaces. If your data is namespaced, though, you could just use `admin-user`, which is different from `user` already (but that would be incoherent, see point 1). Data-wise, you don't need to specify that `user` is _inside_ `admin` because data tends to be flat. If you're conflating your internal code design with how data is designed/shared/expressed, then perhaps you should re-evaluate what you're trying to accomplish.

**Comment by bf4** [comment]
@kurko I agree that many engineering problems can be solved by better design, but some designs are so common, that it's not always sufficient, as an implementor, to say: don't do that.  

I do think that having an implementation-specific namespace -> attribute mapping is a really interesting idea that I've also considered as a recommendation, but haven't every implemented it at a library level.

So, someone has a resource that needs to be processed differently based.  In the implementation, the different handling is determined by the naming of the object or struct.  Is it reasonable to have a record that can be representation as different resources? e.g. `[{ id:1 , type: users, attributes: {name}}]`, `[{ id:1 , type: owners, attributes: {name, title, superpower}, relationships: { ownerships: [{id: 2, type: business}]}]` etc? If so, that's not dissimilar from the case where there's a user v1 and user v2 and an admin-user.

**Comment by kurko** [comment]
@bf4 I didn't understand your analogy 😅

But, I think a good solution for this entire thread is pushing for the implementation of [JSONAPI Extensions](http://jsonapi.org/extensions/), so your organization can benefit from adhoc solutions.

**Comment by jelhan** [comment]
Looking back at this from latest v1.1, I think it could be very well addressed using a [profile](https://jsonapi.org/format/1.1/#profiles), which specifies the _meaning_ of the `type` string. From the perspective of base specification the value of `type` is just an opaque string as pointed out by @ethanresnick. A profile could be used to further specify _what_ that string should mean. It could specify a namespace and a separator to be used.

What that namespace is, shouldn't matter much as long as it is specified by that profile. We are writing APIs to be consumed by programs in the first place. Not read by humans.

As the need of namespaces is not _very_ common, I think it is better addressed by a profile than by a recommendation.

I guess it's time for someone to write such a profile.

**Comment by bf4** [comment]
Closing as undefined and that's ok
